include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -360,11 +364,153 @@ create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_t
+
+      int size_shared;
+      MPI_Comm_size(comm_shared, &size_shared);
+      MPI_Comm_free(&comm_shared);
+
+      // determine maximum, since some shared memory communicators
+      // might not be filed completely
+      return Utilities::MPI::max(size_shared, mpi_comm);
+    }();
+
+    const unsigned int my_rank    = Utilities::MPI::this_mpi_process(mpi_comm);
+    const unsigned int group_root = (my_rank / group_size) * group_size;
+
+    // create coarse grid on root processes
+    Triangulation<dim, spacedim> tria_serial;
+    if(my_rank == group_root)
+    {
+      auto [points, cell_data, sub_cell_data] = GridTools::get_coarse_mesh_description(temp_tria);
+
+      std::vector<std::pair<unsigned int, CellData<dim>>> cell_data_temp;
+
+      unsigned int counter = 0;
+
+      for(const auto & cell : temp_tria.cell_iterators_on_level(0))
+        cell_data_temp.emplace_back(cell->id().get_coarse_cell_id(), cell_data[counter++]);
+
+      std::sort(cell_data_temp.begin(), cell_data_temp.end(), [](const auto & a, const auto & b) {
+        return a.first < b.first;
+      });
+
+      cell_data.clear();
+
+      for(const auto & i : cell_data_temp)
+        cell_data.emplace_back(i.second);
+
+      tria_serial.create_triangulation(points, cell_data, sub_cell_data);
+    }
+
+    // execute refinement on root process
+    if(temp_tria.n_global_levels() > 1)
+    {
+      // collect refinement flags
+      std::vector<std::vector<std::vector<CellId>>> refinement_flags(min_level);
+      {
+        for(unsigned int l = 0; l < min_level; ++l)
+        {
+          std::vector<CellId> local_refinement_flags;
+
+          for(const auto & cell : temp_tria.cell_iterators_on_level(l))
+            if(cell->has_children())
+              local_refinement_flags.push_back(cell->id());
+
+          refinement_flags[l] = Utilities::MPI::gather(mpi_comm, local_refinement_flags, 0);
+        }
+
+        MPI_Comm comm_root;
+        MPI_Comm_split(mpi_comm, my_rank == group_root, my_rank, &comm_root);
+
+        if(my_rank == group_root)
+          refinement_flags = Utilities::MPI::broadcast(comm_root, refinement_flags);
+
+        MPI_Comm_free(&comm_root);
+      }
+
+      // ... actually perform refinement
+      if(my_rank == group_root)
+      {
+        for(unsigned int l = 0; l < refinement_flags.size(); ++l)
+        {
+          unsigned int counter = 0;
+          for(const auto & flags : refinement_flags[l])
+          {
+            for(const auto & cell_id : flags)
+            {
+              tria_serial.create_cell_iterator(cell_id)->set_refine_flag();
+              counter++;
+            }
+          }
+
+          if(counter > 0)
+            tria_serial.execute_coarsening_and_refinement();
+        }
+      }
+    }
+
+    // continue as above but do not work on p:d:T but on the serial
+    // triangulation
+    unsigned int n_partitions_before = Utilities::MPI::n_mpi_processes(MPI_COMM_WORLD);
```

Comment:
I suggest to rename this variable, `before` is not clear. According to the for loop below, `n_partitions_fine` would be more appropriate. Technically, it seems as if we do not even need two variables `n_partitions_before`, `n_partitions`, but that a single variable would be enough?

Moreover, the naming should be in line with the name chosen for `new_tria`, see also my other comment above.